File: android/app/src/main/java/com/HammersTech/RoutineChart/core/data/remote/firebase/CompositeRoutineAssignmentRepository.kt
package com.HammersTech.RoutineChart.core.data.remote.firebase

import com.HammersTech.RoutineChart.core.data.local.repositories.RoomRoutineAssignmentRepository
import com.HammersTech.RoutineChart.core.domain.models.RoutineAssignment
import com.HammersTech.RoutineChart.core.domain.repositories.RoutineAssignmentRepository
import com.HammersTech.RoutineChart.core.utils.AppLogger
import kotlinx.coroutines.flow.Flow
import java.time.Instant
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Composite repository that uses Room as source of truth and syncs assignments to Firestore
 * Phase 3.5: Sync RoutineAssignments (Upload Queue + Pull Cursor)
 */
@Singleton
class CompositeRoutineAssignmentRepository @Inject constructor(
    private val localRepo: RoomRoutineAssignmentRepository,
    private val uploadQueue: AssignmentUploadQueueService,
    private val syncService: FirestoreAssignmentSyncService,
    private val cursorManager: SyncCursorManager
) : RoutineAssignmentRepository {

    override suspend fun create(assignment: RoutineAssignment) {
        localRepo.create(assignment)
        AppLogger.Database.info("Created assignment locally (will sync via upload queue): ${assignment.id}")
    }

    override suspend fun update(assignment: RoutineAssignment) {
        localRepo.update(assignment)
        AppLogger.Database.info("Updated assignment locally (will sync via upload queue): ${assignment.id}")
    }

    override suspend fun getById(id: String): RoutineAssignment? =
        localRepo.getById(id)

    override suspend fun getByFamilyId(familyId: String): List<RoutineAssignment> =
        localRepo.getByFamilyId(familyId)

    override suspend fun getActiveByChildId(childId: String): List<RoutineAssignment> =
        localRepo.getActiveByChildId(childId)

    override fun observeActiveByChildId(childId: String): Flow<List<RoutineAssignment>> =
        localRepo.observeActiveByChildId(childId)

    override suspend fun getByRoutineId(routineId: String): List<RoutineAssignment> =
        localRepo.getByRoutineId(routineId)

    override suspend fun softDelete(id: String) {
        localRepo.softDelete(id)
        AppLogger.Database.info("Soft deleted assignment locally (will sync via upload queue): $id")
    }

    /** Upload all unsynced assignments for a family. Returns count uploaded. */
    suspend fun uploadUnsynced(familyId: String): Int =
        uploadQueue.uploadUnsyncedAssignments(familyId)

    /** Pull assignments from Firestore updated since last sync. Last-write-wins by updatedAt. Returns count merged. */
    suspend fun pullAssignments(familyId: String): Int {
        AppLogger.Database.info("🔄 Starting pull of assignments from Firestore for familyId: $familyId")

        val cursorCollection = "routine_assignments"
        val cursor = cursorManager.getCursor(cursorCollection)
        val lastSyncedAt = cursor?.lastSyncedAt ?: Instant.ofEpochSecond(0)

        val remoteAssignments = try {
            syncService.getAssignmentsUpdatedSince(familyId, lastSyncedAt)
        } catch (e: Exception) {
            AppLogger.Database.error("❌ Failed to query Firestore for assignments: ${e.message}", e)
            cursorManager.updateCursor(cursorCollection, Instant.now())
            return 0
        }

        if (remoteAssignments.isEmpty()) {
            AppLogger.Database.info("✅ No new assignments to pull from Firestore")
            cursorManager.updateCursor(cursorCollection, Instant.now())
            return 0
        }

        AppLogger.Database.info("📥 Found ${remoteAssignments.size} assignment(s) to pull from Firestore")

        var mergedCount = 0
        var skippedCount = 0

        remoteAssignments.forEach { remote ->
            try {
                val local = localRepo.getById(remote.id)
                if (local != null) {
                    if (remote.updatedAt.isAfter(local.updatedAt)) {
                        localRepo.saveFromFirestore(remote)
                        mergedCount++
                        AppLogger.Database.info("✅ Merged assignment (remote wins): ${remote.id}")
                    } else {
                        skippedCount++
                        AppLogger.Database.info("⏭️ Skipped assignment (local wins): ${remote.id}")
                    }
                } else {
                    localRepo.saveFromFirestore(remote)
                    mergedCount++
                    AppLogger.Database.info("✅ Inserted new assignment from Firestore: ${remote.id}")
                }
            } catch (e: Exception) {
                AppLogger.Database.error("❌ Failed to merge assignment ${remote.id}: ${e.message}", e)
            }
        }

        cursorManager.updateCursor(cursorCollection, Instant.now())
        AppLogger.Database.info("✅ Pull complete: merged $mergedCount assignment(s), skipped $skippedCount")
        return mergedCount
    }
}

File: ios/RoutineChart/RoutineChart/Core/Data/Remote/Firebase/CompositeRoutineAssignmentRepository.swift
//
//  CompositeRoutineAssignmentRepository.swift
//  RoutineChart
//
//  Phase 3.5: Composite repository - SQLite (source of truth) + Firestore sync via upload queue and pull cursor
//

import Foundation
import GRDB
import OSLog

/// Composite repository that uses SQLite as source of truth and syncs assignments to Firestore via upload queue and pull cursor
final class CompositeRoutineAssignmentRepository: RoutineAssignmentRepository {
    private let localRepo: SQLiteRoutineAssignmentRepository
    private let uploadQueue: AssignmentUploadQueueService
    private let syncService: FirestoreAssignmentSyncService
    private let cursorManager: SyncCursorManager
    
    init(
        localRepo: SQLiteRoutineAssignmentRepository = SQLiteRoutineAssignmentRepository(),
        uploadQueue: AssignmentUploadQueueService = AssignmentUploadQueueService(),
        syncService: FirestoreAssignmentSyncService = FirestoreAssignmentSyncService(),
        cursorManager: SyncCursorManager = SyncCursorManager()
    ) {
        self.localRepo = localRepo
        self.uploadQueue = uploadQueue
        self.syncService = syncService
        self.cursorManager = cursorManager
    }
    
    // MARK: - Repository Methods (Local-First)
    
    func create(_ assignment: RoutineAssignment) async throws {
        try await localRepo.create(assignment)
        AppLogger.database.info("Created assignment locally (will sync via upload queue): \(assignment.id)")
    }
    
    func get(id: String) async throws -> RoutineAssignment? {
        try await localRepo.get(id: id)
    }
    
    func update(_ assignment: RoutineAssignment) async throws {
        try await localRepo.update(assignment)
        AppLogger.database.info("Updated assignment locally (will sync via upload queue): \(assignment.id)")
    }
    
    func getAll(familyId: String) async throws -> [RoutineAssignment] {
        try await localRepo.getAll(familyId: familyId)
    }
    
    func getByChild(familyId: String, childId: String) async throws -> [RoutineAssignment] {
        try await localRepo.getByChild(familyId: familyId, childId: childId)
    }
    
    func getByRoutine(familyId: String, routineId: String) async throws -> [RoutineAssignment] {
        try await localRepo.getByRoutine(familyId: familyId, routineId: routineId)
    }
    
    func softDelete(id: String) async throws {
        try await localRepo.softDelete(id: id)
        AppLogger.database.info("Soft deleted assignment locally (will sync via upload queue): \(id)")
    }
    
    // MARK: - Upload Queue (Phase 3.5)
    
    /// Upload all unsynced assignments for a family
    func uploadUnsynced(familyId: String) async throws -> Int {
        try await uploadQueue.uploadUnsyncedAssignments(familyId: familyId)
    }
    
    // MARK: - Pull Cursor (Phase 3.5)
    
    /// Pull assignments from Firestore that were updated since the last sync
    /// Applies merge logic (last-write-wins by updatedAt) and updates the sync cursor
    func pullAssignments(familyId: String) async throws -> Int {
        AppLogger.database.info("🔄 Starting pull of assignments from Firestore for familyId: \(familyId)")
        
        let cursor = try await cursorManager.getCursor(collection: "routine_assignments")
        let lastSyncedAt = cursor?.lastSyncedAt ?? Date(timeIntervalSince1970: 0)
        
        let remoteAssignments = try await syncService.getAssignmentsUpdatedSince(familyId: familyId, since: lastSyncedAt)
        
        guard !remoteAssignments.isEmpty else {
            AppLogger.database.info("✅ No new assignments to pull from Firestore")
            try await cursorManager.updateCursor(collection: "routine_assignments", lastSyncedAt: Date())
            return 0
        }
        
        AppLogger.database.info("📥 Found \(remoteAssignments.count) assignment(s) to pull from Firestore")
        
        var mergedCount = 0
        var skippedCount = 0
        
        for remote in remoteAssignments {
            do {
                if let local = try await localRepo.get(id: remote.id) {
                    if remote.updatedAt > local.updatedAt {
                        try await saveAssignmentFromFirestore(remote)
                        mergedCount += 1
                        AppLogger.database.info("✅ Merged assignment (remote wins): \(remote.id)")
                    } else {
                        skippedCount += 1
                        AppLogger.database.info("⏭️ Skipped assignment (local wins): \(remote.id)")
                    }
                } else {
                    try await saveAssignmentFromFirestore(remote)
                    mergedCount += 1
                    AppLogger.database.info("✅ Inserted new assignment from Firestore: \(remote.id)")
                }
            } catch {
                AppLogger.database.error("❌ Failed to merge assignment \(remote.id): \(error.localizedDescription)")
            }
        }
        
        try await cursorManager.updateCursor(collection: "routine_assignments", lastSyncedAt: Date())
        AppLogger.database.info("✅ Pull complete: merged \(mergedCount) assignment(s), skipped \(skippedCount)")
        return mergedCount
    }
    
    /// Save assignment from Firestore to local database and mark as synced
    private func saveAssignmentFromFirestore(_ assignment: RoutineAssignment) async throws {
        let db = try SQLiteManager.shared.database()
        try await db.write { db in
            var mutable = assignment
            if try RoutineAssignment.exists(db, key: assignment.id) {
                try mutable.update(db)
            } else {
                try mutable.insert(db)
            }
            try db.execute(sql: "UPDATE routine_assignments SET synced = 1 WHERE id = ?", arguments: [assignment.id])
        }
    }
}
